kontonummer.js

```diff
@@ -367,12 +367,22 @@
      * @returns {Object|Boolean}
      */
     var kontonummer = function(number) {
-        var n = number.replace(/\D/g, ''), i, bank, errors = [];
-        var matchedBanks = [];
-        var isValid = false;
+        var accountNumber = number || false
+
+        if (!accountNumber
+          || !/^\d+$/.test(accountNumber)
+          || accountNumber.constructor === Number) {
```

Comment:
### typeof

`typeof` really depends on what we're expecting, in this case it could work. Though it's good at differentiating between Primitives it's not very good at diff'ing different objects.

`typeof`
```js
typeof 3 // "number"
typeof "abc" // "string"
typeof {} // "object"
typeof true // "boolean"
typeof undefined // "undefined"
typeof function() {} // "function"

// And where it falls short..
typeof [] // "object"
typeof null // "object"
typeof new Date() // "object"
typeof new Regex() // "object"
```

### instanceof

Meanwhile `instanceof` tells you whether an object is an instance of a certain type. In this case the `type` here is the constructor, still.

```js
function Animal(){}
var a = new Animal()
a instanceof Animal // true

// And checking against the property `constructor`
a.constructor === Animal // true
```

Now the difference here is that the `.constructor`-type check doesn't walk up the `prototype` chain (which `instanceof` does and performance wise is worse).

Also running the type check using the `constructor` property will throw `TypeError` if the object you're checking is either `null` or `undefined`. Which could be nice to have at some cases where you want to make sure what you get is right otherwise exit out entirely.

Now looking at how `instanceof` deals with certain types.

`instanceof`

```js
[1, 2, 3] instanceof Array // true
/abc/ instanceof RegExp // true
({}) instanceof Object // true
(function() {}) instanceof Function // true

// but then..
3 instanceof Number // false
true instanceof Boolean // false
'abc' instanceof String // false

// "null" or "undefined" whcih are not an instance of anything
null instanceof Boolean // false
undefined instanceof Array // false
```

Here we see that `instanceof` doesn't work well with Primitives.

### constructor type checking

Now what if we use the `constructor´ property to check the same types.

`constructor`

```js
(3).constructor === Number // true
true.constructor === Boolean // true
'abc'.constructor === String // true
```

They result is now correct, the magic here is what's going on behind the scenes. Accessing the property of a Primitive object javascript will wrap it with an Object wrapper leaving us with

```js
// The wrapping is nothing we see, it's done behind the scenes
new Number(3) instanceof Number // true
new Boolean(true) instanceof Boolean // true
new String('abc') instanceof String // true
```

So in the end I guess it depends on what we want, but I'm more leaning towards the `constructor` method as the result is more precise, though it has its pitfalls such as being able to check against the parent of a deriving class, the `TypeError` it will throw if you try accessing it on `null` and `undefined` and lastly it will not walk up the `prototype` chain.

Reply:
I'm aware of the `typeof` having shortcomings in regards to objects, but for checking primitive types like number or string that want here, it should be fine. I'm not sure what I was thinking about regarding `instanceof`, I agree it's probably the worst of the three in this case.

As long as we keep the `!accountNumber` check first, using `accountNumber.constructor` will be fine also with `null` and `undefined`.